semantics/src/traversal/expressions.rs

```diff
@@ -323,6 +323,32 @@ fn expr_comp_operation(
     unreachable!()
 }
 
+fn expr_ternary(
+    scope: Shared<BlockScope>,
+    context: Shared<Context>,
+    exp: &Spanned<fe::Expr>,
+) -> Result<ExpressionAttributes, SemanticError> {
+    if let fe::Expr::Ternary {
+        test,
+        if_expr,
+        else_expr,
+    } = &exp.node
+    {
+        let test_attributes = expr(Rc::clone(&scope), Rc::clone(&context), test);
```

Comment:
One thing we need to check here is that the `test` is actually a boolean expression. Similar to

https://github.com/ethereum/fe/blob/c2681b844bf441ee296356ea6ab41f86fbd1e125/semantics/src/traversal/functions.rs#L232

Reply:
ahh, the hardcoded boolean statement makes me confuse. No worries I will make that change.

FYI here we need to check the `if_attributes` as a boolean expression, not the `test` as the test consists of the successful return value. I think we need to work on the naming or make some re-arrangements if we are expecting to `test` as the boolean expression. 
Here is the declaration of the ternary expression https://github.com/ethereum/fe/blob/c2681b844bf441ee296356ea6ab41f86fbd1e125/parser/src/ast.rs#L196

Reply:
Not sure I follow, let's break it down:

- `test`: This is the expression that is being tested e.g. `foo` in `1 if foo else 0`. The thing that we need to check for is that `foo` would would be an expression that evaluates to a `boolean` because non-booleans are illegal as a `test` in a ternary operator. E.g. we can not allow `foo` if it is of type `u256`.

- `if_expr`: The expression that is evaluated if `test` is `true`
- `else_expr`: The expression that is evaluated if `test` is false.

The thing that we need to check is that `if_expr` and `else_expr` evaluate to the same type. E.g. we can not allow one side to evaluate to a `address` if the other side evaluates to a `u256`

Reply:
I think this is the syntax we are trying to achieve as per Python grammar:
```
[on_true] if [expression] else [on_false]
```
So for our example

```
contract Foo:

    pub def bar(input: u256) -> u256:
        return 1 if input > 5 else 0
```

on_true = 1, expression = input > 5, on_false = 0

So by the parser, I got the Ternary expression values something like this - 

 `test` - Spanned { node: Num("1"), span: Span { start: 68, end: 69 } } which lead to `test_attribute` value is ExpressionAttributes { typ: Base(U256), location: Value }, So I don't think it will ever be boolean it will always be the evaluated value if `if_expression` is true.

`if_expr` - Spanned { node: CompOperation { left: Spanned { node: Name("input"), span: Span { start: 73, end: 78 } }, op: Spanned { node: Gt, span: Span { start: 79, end: 80 } }, right: Spanned { node: Num("5"), span: Span { start: 81, end: 82 } } }, span: Span { start: 73, end: 82 } } which lead to `if_expr_attribute` value is ExpressionAttributes { typ: Base(Bool), location: Value }. Which always be the boolean and it should be.

`else_expr` - Spanned { node: Num("0"), span: Span { start: 88, end: 89 } } which lead to `else_attribute` value is ExpressionAttributes { typ: Base(U256), location: Value } . This is expected result.

So what I mean above is as per your terminology `test` & `if_expr` value should be used interchangebly but as per the code, it is reverse or in other words, it is not as per your terminology you above mention. So what I suggest above either we change the Ternary declaration from 
```
Ternary {
        test: Box<Spanned<Expr<'a>>>,
        if_expr: Box<Spanned<Expr<'a>>>,
        else_expr: Box<Spanned<Expr<'a>>>,
    },  
to 
Ternary {
        if_expr: Box<Spanned<Expr<'a>>>,
        test: Box<Spanned<Expr<'a>>>,
        else_expr: Box<Spanned<Expr<'a>>>,
    },  
```
Or we can fix the parser.

BTW I already in place the check for the same type that is evaluated from the `if_expr` & `else_expr`.

Reply:
>to 
```
Ternary {
        if_expr: Box<Spanned<Expr<'a>>>,
        test: Box<Spanned<Expr<'a>>>,
        else_expr: Box<Spanned<Expr<'a>>>,
    },
```

Ah, you mean changing the order of the enum properties to match the syntax of the language. We can definitely do that :+1:

Reply:
Minor parser changes were needed to facilitate the ordering.